### DISCUSSION TITLE
How to enable Typst rendering of math

### DISCUSSION THREAD

**Discussion by sverrirarnors** [quarto-dev/quarto-cli#11368] [Q&A] [answered]
### Description

Currently, math content inside `$` marks renders as Latex through pandoc (see https://github.com/quarto-dev/quarto-cli/issues/6560). Is there a way to let Typst handle the rendering instead? Currently I solve this by using raw Typst blocks, but that is a bit cumbersome in the long-run.

**Comment by cderv** [comment]
> Currently, math content inside $ marks renders as Latex through pandoc

What do you mean by this ? Example from #6560 is 
````markdown
---
format: typst
keep-typ: true
---

## Math

For more information, see @eq-proof.

$$
\frac{1}{2} \equiv 1 \div 2
$$ {#eq-proof}
````

and renders to this typst
````typst
#math.equation(block: true, numbering: "(1)", [ $ 1 / 2 equiv 1 div 2 $ ])<eq-proof>
````

This is Typst code than it written by Pandoc (for the equation part) and Quarto (for the reference label part on the block). It uses the dedicated equation syntax: https://typst.app/docs/reference/math/equation/#syntax

> Is there a way to let Typst handle the rendering instead? Currently I solve this by using raw Typst blocks, but that is a bit cumbersome in the long-run.

Can you share an example on what you are referring too ? What is not working ? Or how you think this is working ? Note that Pandoc conversion is involved when not using Raw content, and the common math syntax for Pandoc's Markdown is using `$` for inline math, and `$$` for block math, and then pandoc will convert to math syntax for the output document.  

Related: There was a question once about using `mitex` Typst package in Quarto: https://github.com/quarto-dev/quarto-cli/discussions/11075

**Comment by sverrirarnors** [comment]
Thanks for the reply! I guess it's more of a Pandoc issue then, if the math mode in Pandoc uses Latex as the markup.

What I would like to do is use Typst as the markup language for math mode instead, so that I could write Typst code inside of the `$$` tags instead of Latex.

For example, this `$underbrace(e^(i pi), -1) + 1 = 0$` currently evaluates to 
![image](https://github.com/user-attachments/assets/a7f281af-4d14-41f7-b2c5-9ad49681f5fb),

but in Typst would evaluate to 
![image](https://github.com/user-attachments/assets/ae9fb5df-0448-4768-b1f8-6bb545e22a1b)

**Reply by mcanouil** [reply]
Use raw block content then

- https://quarto.org/docs/authoring/markdown-basics.html#raw-content

**Reply by cderv** [reply, marked as answer]
Yes this is how Math work in Pandoc. More at https://pandoc.org/MANUAL.html#math

If you want to write raw typst, then you need raw block as you already do. 

You could also catched the math in a lua filter, and do output it as a raw block directly. 

````lua

-- Inline Math
Math = function(m)
  if not quarto.doc.is_format('typst') then
    return nil
  end
  if m.mathtype == "InlineMath" then
    return pandoc.RawInline('typst', '$'.. m.text .. '$')
  end
end

-- DisplayMath
-- We want to replace all Para by a RawBlock
Para = function(p)
  if not quarto.doc.is_format('typst') then
    return nil
  end
  if p.content[1].t == "Math" and p.content[1].mathtype == "DisplayMath" then
    return pandoc.RawBlock('typst', '$' .. p.content[1].text .. '$')
  end
end


````

and using this filters in 
````yaml
---
title: "Quarto Playground"
format: typst
filters:
  - test.lua
keep-typ: true
---

This is an equation: $underbrace(e^(i pi), -1) + 1 = 0$

$$
underbrace(e^(i pi), -1) + 1 = 0
$$

````

![image](https://github.com/user-attachments/assets/578989aa-d670-41e3-9dc4-d533d38dcef0)

Probably with more adjustments to handle more cases. 

I don't know if this is can safely be a built-in feature. Maybe there should be a Pandoc option that would allow to use typst match directly in typst format, with no TeX to Typst conversion. You could ask on Pandoc issue / Discussion board about this. 

Hope it helps

**Reply by sverrirarnors** [reply]
Thanks @cderv! This works perfectly for my use case.

I'll also post this in the Pandoc discussion board – probably most appropriate to solve it there.

**Reply by arnaudgallou** [reply]
@cderv Any chance to implement this in Quarto and toggle it with a YAML option?

**Reply by cderv** [reply]
I think it would be best as a feature request to Pandoc to allow math to be treated as Raw typst math. This way it would benefit all pandoc users, and we could inherit it. 

If they don't want that, we could consider something. However, I think it still fit nicely in an extension.
 
The problem for Quarto is that for main feature, we want to have markdown syntax that is common across format, and with this math logic, this is no more a Math syntax that would work in any other format than typst. 

Usually for format specific usage, raw code is the way to go. Lua filter is there by convenience, it can work without 
````markdown
---
title: "Quarto Playground"
format: typst
keep-typ: true
---

This is an equation: `$underbrace(e^(i pi), -1) + 1 = 0$`{=typst}

```{=typst}
$underbrace(e^(i pi), -1) + 1 = 0$
```
````

So probably keeping in a extension is a good fit - something called `typst-math` maybe with the other extension about mitex. 🤔

**Reply by sverrirarnors** [reply]
I asked this in the Pandoc discussion and I think it won't be implemented there soon (https://github.com/jgm/pandoc/discussions/10381#discussioncomment-11231887), unless more people call for it.

But I would also definitely use an extension that would allow for this!

**Reply by mcanouil** [reply]
You already have all the pieces, all you need to do is follow https://quarto.org/docs/extensions/filters.html or https://quarto.org/docs/extensions/formats.html

**Reply by sverrirarnors** [reply]
Thanks, @mcanouil. Working on this now.

**Reply by sverrirarnors** [reply]
Here's an extension: https://github.com/sverrirarnors/typst-math.

Thanks for the suggestion @mcanouil and for the code @cderv!

**Reply by mcanouil** [reply]
FYI, if you want you could also make a format extension which has the filter enable by default.

And if you work with multiple formats, you might want to make it clear from the YAML front matter that the filter is only for Typst.
```yaml
format:
  typst:
    filters:
      - your-filter.lua
```

**Reply by VarLad** [reply]
Hello! 
@sverrirarnors Any idea if the same can be used to render typst in word documents as well?

**Reply by sverrirarnors** [reply]
Probably not, because the lua sidesteps pandoc's handling of math, and would probably just result in the raw typst code being output. This only works on PDFs if you use typst as an intermediary step, but as far as I know typst is not used at all for word rendering.

I could rewrite my plugin to use [this workaround ](https://fosstodon.org/@pandoc/112926076691256397) instead of the raw typst blocks, and then it would likely work for word rendering. But the typst to latex conversion is not perfect so I'd probably still use the raw block method because I primarily render in PDF.

Ultimately, this is a pandoc issue, so if you want to raise this further you can advocate for it here: https://github.com/jgm/pandoc/discussions/10381
